ast/compile.go

```diff
@@ -1353,6 +1354,31 @@ func (c *Compiler) getExports() *util.HashMap {
 	return rules
 }
 
+func (c *Compiler) checkUnusedImports() {
+	if !c.strict {
+		return
+	}
+
+	for _, name := range c.sorted {
+		mod := c.Modules[name]
+
+		for _, imp := range mod.Imports {
+			name := imp.Name()
+			processedImports := make(map[string]string)
+
+			for _, r1 := range mod.Rules {
+				if strings.Contains(r1.Body.String(), imp.String()[6:]) || strings.Contains(r1.Body.String(), name.String()) {
+					processedImports[name.String()] = imp.String()
+				}
```

Comment:
Oh that's an interesting approach. Took me a moment to understand 😃 

I think we'd be better off using one of the `Walk*` methods here (`ast.WalkTerms` or `ast.WalkExprs` or whatever fits); because right now, this check is prone to false positives:

```rego
package p

import data.foo.bar as bar

r {
  input.bar == 2
}
```

Reply:
Sorry if false positive was a the wrong word -- I mean that when the compiler runs this stage, it should be clear if we're using the import, *unambiguously*.  So, neither `input.bar` nor `hello.bar` should not make the compiler think that `import data.foo.bar as bar` (or a plain `import data.foo.bar`) was used.

I often find myself adding `log.Printf("%s: %v", s.name, c.Modules)` to [this function](https://github.com/open-policy-agent/opa/blob/6dde17fd403831a2cbd9f8e285d3972b57b5b02d/ast/compile.go#L1274) to figure out what the compiler does in rewriting. With a module like the example, we'll find
```
2022/02/28 08:25:43 CheckDuplicateImports: map[p.rego:package p

import data.foo.bar as bar

r = true { equal(input.bar, 2) }]
2022/02/28 08:25:43 CheckKeywordOverrides: map[p.rego:package p

import data.foo.bar as bar

r = true { equal(input.bar, 2) }]
2022/02/28 08:25:43 ResolveRefs: map[p.rego:package p

r = true { equal(input.bar, 2) }]
```

So `ResolveRefs` has removed the import, and didn't otherwise change the module.

With a positive example, 
```
package p

import data.foo.bar

r {
  input.bar == bar
}
```
we get this:
```
2022/02/28 08:27:29 CheckDuplicateImports: map[q.rego:package p

import data.foo.bar

r = true { equal(input.bar, bar) }]
2022/02/28 08:27:29 CheckKeywordOverrides: map[q.rego:package p

import data.foo.bar

r = true { equal(input.bar, bar) }]
2022/02/28 08:27:29 ResolveRefs: map[q.rego:package p

r = true { equal(input.bar, data.foo.bar) }]
```

So we should be able to find unused imports by 
- for all imports
   - take the ref of the import (qualified name won't matter because it's been rewritten already)
   - check if the ref is used in the body
   - fail if it isn't
 
This is merely a sketch, and there may be better ways to approach finding unused imports here.

Concretely, I think what I've outlined would fail with this example:

```
package p
import data.foo # unused
r { data.foo == 10 }
```
because at the stage we'd check for used imports, the import would already have been removed, but it didn't actually resolve any ref.

This makes me think that perhaps finding unused imports should be a stage that closely mirrors what ResolveRefs does, instead of running _after ResolveRefs_. What do you think?

(Btw these would all make decent test cases when you get to adding a few test cases. Personally, I prefer to start with them, because it gives me some assurance when expanding the functionality that previous wins aren't lost again 😅)

Reply:
For sure. I understand your concept. I uploaded some new code that is passing all of the test cases I could come up with. Please check it out and let me know what you think.